_posts/2021-09-07-what-javascript-tests-could-learn-from-rspec.md

```diff
@@ -0,0 +1,227 @@
+
+The difference is in, well, context. Well-organized RSpec tests use `describe` to describe what's being
+tested...and `context` to describe scenarios of the thing being tested.
+
+For example, if I wanted to test the `multiply` method of a `Calculator` class, I might write some test scenarios
+that look like this:
+
+```rb
+describe "Calculator" do
+  describe ".multiply" do
+    context "when the first value is negative" do
+      context "when the second value is negative" do
+        it "returns a positive number" do
+        end
+      end
+      context "when the second value is positive" do
+        it "returns a negative number" do
+        end
+      end
+    end
+  end
+end
+```
+
+See the difference in those test cases between `describe` and `context`? The way I think about it is: if the
+statement coming after my `describe`/`context` describes a pre-condition for the test, it's a `context`; otherwise
+it's a `describe`.
+
+## What's the big deal?
+
+On the surface these two features don't seem like much, but they provide a really powerful framework for organizing
+test cases and the associated test setup.
+
+With the `let` blocks being lazily evaluated, and override-able, I can set up data at the exact level of tests that
+I need it. When I need to override it for a certain set of tests, I can put another `let` block in that set. In
+JavaScript I can define functions to set up my data for me with just the changes I need at each test level, or I
+can use `beforeEach` blocks, but all that can get pretty noisy.
+
+And with `context` blocks, I can more clearly lay out the scenarios of my tests. Yes, I _could_ just do this with
+more `describe` blocks in JavaScript, but how often have you found JavaScript tests that actually do this? I've
+personally seen/written too many tests to count named something like
+`it("returns false when the flag is enabled, they're located in the US, but they have brown hair.")`. That's three
+scenarios rolled into one test name. It works, but being able to nest different `context` blocks to define my
+complex scenarios is much easier to read.
+
+## Examples
+
+Here's an example of some tests I could write in RSpec with `let` and `context`:
+
+```rb
+describe "Calculator" do
+  let(:calculator) { Calculator.new }
+
+  describe ".multiply" do
+    let(:result) { calculator.multiply(first, second) }
+
+    context "when the first value is negative" do
+      let(:first) { -1 }
+
+      context "when the second value is negative" do
+        let(:second) { -3 }
+
+        it "returns a positive number" do
+          result.should eq(3)
+        end
+      end
+
+      context "when the second value is positive" do
+        let(:second) { 3 }
+
+        it "returns a negative number" do
+          result.should eq(-3)
+        end
+      end
+    end
+  end
+end
+```
+
+HOW COOL IS THAT! Every `describe`/`context` block has _exactly_ the setup data it needs defined clearly inside it.
```

Comment:
Love the enthusiasm - it's fun to read a post when the author is excited about the subject!